### ISSUE TITLE
[Flang][OpenMP] Internal Compiler Error with atomic update

### ISSUE THREAD

**Issue by kiranchandramohan** [llvm/llvm-project#61089]
The OpenMP atomic update codegen in Flang inserts the load/creation of the value with which to update also into the atomic region. This triggers an ICE in the translation from OpenMP+LLVM MLIR dialect to LLVM IR. The translation expects only two operations in the atomic update region:
1. The update operation
2. The yield
I guess the strong requirement is due to the conversion of the update operation to an atomicrmw instruction is possible. So the update operation has to be identifiable.

Flang patch : https://reviews.llvm.org/D125668
MLIR patch : https://reviews.llvm.org/D119657

**Reproducer**

```f90
subroutine test()
        real :: x
        real :: y
        x = 20
        y = 10
        !$OMP ATOMIC
        y = x + y
end subroutine test
```

**Comment by NimishMishra** [comment]
Hi @kiranchandramohan 
(Adding @kiranktp also).

I believe we need to modify the LLVM IR verifier in this case. The design taken for `atomic update` constructs a basic block within which arbitrary number of operations can reside. 

Is it ok with you?


```
llvm.func @_QPtest() {                                                                                                                       
  %0 = llvm.mlir.constant(1.000000e+01 : f32) : f32                                                                                         
  %1 = llvm.mlir.constant(2.000000e+01 : f32) : f32                                                                                         
  %2 = llvm.mlir.constant(1 : i64) : i64                                                                                                     
  %3 = llvm.alloca %2 x f32 {bindc_name = "x", in_type = f32, operand_segment_sizes = array<i32: 0, 0>, uniq_name = "_QFtestEx"} : (i64) -> !
llvm.ptr<f32>                                                                                                                               
  %4 = llvm.alloca %2 x f32 {bindc_name = "y", in_type = f32, operand_segment_sizes = array<i32: 0, 0>, uniq_name = "_QFtestEy"} : (i64) -> !
llvm.ptr<f32>                                                                                                                               
  llvm.store %1, %3 : !llvm.ptr<f32>                                                                                                         
  llvm.store %0, %4 : !llvm.ptr<f32>                                                                                                         
  omp.atomic.update   %4 : !llvm.ptr<f32> {
  ^bb0(%arg0: f32):
    %5 = llvm.load %3 : !llvm.ptr<f32>
    %6 = llvm.fadd %5, %arg0  {fastmathFlags = #llvm.fastmath<contract>} : f32
    omp.yield(%6 : f32)
  }
  llvm.return
```

**Comment by NimishMishra** [comment]
If I disable the checks, the following IR is generated.

```
define void @_QPtest() {
  %x.new.val = alloca float, align 4
  %1 = alloca float, i64 1, align 4
  %2 = alloca float, i64 1, align 4
  store float 2.000000e+01, ptr %1, align 4
  store float 1.000000e+01, ptr %2, align 4
  br label %entry

entry:                                            ; preds = %0
  %.atomic.load = load atomic i32, ptr %2 monotonic, align 4
  br label %.atomic.cont

.atomic.cont:                                     ; preds = %.atomic.cont, %entry
  %3 = phi i32 [ %.atomic.load, %entry ], [ %8, %.atomic.cont ]
  %.atomic.fltCast = bitcast i32 %3 to float
  %4 = load float, ptr %1, align 4
  %5 = fadd contract float %4, %.atomic.fltCast
  store float %5, ptr %x.new.val, align 4
  %6 = load i32, ptr %x.new.val, align 4
  %7 = cmpxchg ptr %2, i32 %3, i32 %6 monotonic monotonic, align 4
  %8 = extractvalue { i32, i1 } %7, 0
  %9 = extractvalue { i32, i1 } %7, 1
  br i1 %9, label %.atomic.exit, label %.atomic.cont

.atomic.exit:                                     ; preds = %.atomic.cont
  ret void
}
```

**Comment by NimishMishra** [comment]
@kiranchandramohan Can you take a look at my understanding on this issue once you are free? Then I can proceed to fix this issue.

**Comment by kiranchandramohan** [comment]
I think the generated code looks fine.

The fix should be based on a condition check:
if the region has only two operations (the update and the yield) then try the atomic operation generation. If the atomic generation is not possible then generate `cmpxchg` as usual.
If the region has more than two operations then generate `cmpxchg`.

Also, as future work, we should try to run passes that can hoist these extra operations out of the atomic region.
